File: library/kotlin/src/io/envoyproxy/envoymobile/ResponseBuilder.kt
package io.envoyproxy.envoymobile

/**
 * Builder used for constructing instances of `Response` types.
 *
 */
class ResponseBuilder {

  // Headers to send with the response.
  // Multiple values for a given name are valid, and will be sent as comma-separated values.
  private val headers: MutableMap<String, MutableList<String>> = mutableMapOf()

  // Trailers to send with the response.
  // Multiple values for a given name are valid, and will be sent as comma-separated values.
  private val trailers: MutableMap<String, MutableList<String>> = mutableMapOf()

  // Serialized data of the response.
  private var body: ByteArray? = null

  private var status: Int = 200

  /**
   * Add an http status code for the response
   *
   * @param status the http status code
   * @return this builder
   */
  fun addStatus(status: Int): ResponseBuilder {
    this.status = status
    return this
  }

  /**
   * Add serialized data of the response.
   *
   * @param body serialized data of the response
   * @return this builder
   */
  fun addBody(body: ByteArray?): ResponseBuilder {
    this.body = body
    return this
  }

  /**
   * Append a value to the header key.
   *
   * @param name the header key.
   * @param value the value associated to the header key.
   * @return this builder.
   */
  fun addHeader(name: String, value: String): ResponseBuilder {
    if (headers.containsKey(name)) {
      headers[name]!!.add(value)
    } else {
      headers[name] = mutableListOf(value)
    }
    return this
  }

  /**
   * Remove the value in the specified header.
   *
   * @param name the header key to remove.
   * @param value the value to be removed.
   * @return this builder.
   */
  fun removeHeader(name: String, value: String): ResponseBuilder {
    if (headers.containsKey(name)) {
      headers[name]!!.remove(value)
      if (headers[name]!!.isEmpty()) {
        headers.remove(name)
      }
    }
    return this
  }

  /**
   * Remove all headers with this name.
   *
   * @param name the header key to remove.
   * @return this builder.
   */
  fun removeHeaders(name: String): ResponseBuilder {
    headers.remove(name)
    return this
  }

  /**
   * Append a value to the trailer key.
   *
   * @param name the trailer key.
   * @param value the value associated to the trailer key.
   * @return this builder.
   */
  fun addTrailer(name: String, value: String): ResponseBuilder {
    if (trailers.containsKey(name)) {
      trailers[name]!!.add(value)
    } else {
      trailers[name] = mutableListOf(value)
    }
    return this
  }

  /**
   * Remove the value in the specified trailer.
   *
   * @param name the trailer key to remove.
   * @param value the value to be removed.
   * @return this builder.
   */
  fun removeTrailers(name: String): ResponseBuilder {
    trailers.remove(name)
    return this
  }

  /**
   * Remove the value in the specified trailer.
   *
   * @param name the trailer key to remove.
   * @param value the value to be removed.
   * @return this builder.
   */
  fun removeTrailer(name: String, value: String): ResponseBuilder {
    if (trailers.containsKey(name)) {
      trailers[name]!!.remove(value)

      if (trailers[name]!!.isEmpty()) {
        trailers.remove(name)
      }
    }
    return this
  }
}

File: library/swift/src/ResponseBuilder.swift
import Foundation

/// Builder used for constructing instances of `Response` types.
@objcMembers
public final class ResponseBuilder: NSObject {
  /// Status code returned with the response.
  public private(set) var statusCode: Int = 200
  /// Headers returned with the response.
  /// Multiple values for a given name are valid, and will be sent as comma-separated values.
  public private(set) var headers: [String: [String]] = [:]
  /// Trailers returned with the response.
  /// Multiple values for a given name are valid, and will be sent as comma-separated values.
  public private(set) var trailers: [String: [String]] = [:]
  /// Serialized data returned as the body of the response.
  public private(set) var body: Data?

  // MARK: - Initializers

  /// Internal initializer used for converting a response back to a builder.
  convenience init(response: Response) {
    self.init()
    self.statusCode = response.statusCode
    self.headers = response.headers
    self.trailers = response.trailers
    self.body = response.body
  }

  // MARK: - Builder functions

  @discardableResult
  public func addStatusCode(_ statusCode: Int) -> ResponseBuilder {
    self.statusCode = statusCode
    return self
  }

  @discardableResult
  public func addHeader(name: String, value: String) -> ResponseBuilder {
    self.headers[name, default: []].append(value)
    return self
  }

  @discardableResult
  public func removeHeaders(name: String) -> ResponseBuilder {
    self.headers.removeValue(forKey: name)
    return self
  }

  @discardableResult
  public func removeHeader(name: String, value: String) -> ResponseBuilder {
    self.headers[name]?.removeAll(where: { $0 == value })
    if self.headers[name]?.isEmpty == true {
      self.headers.removeValue(forKey: name)
    }

    return self
  }

  @discardableResult
  public func addTrailer(name: String, value: String) -> ResponseBuilder {
    self.trailers[name, default: []].append(value)
    return self
  }

  @discardableResult
  public func removeTrailer(name: String) -> ResponseBuilder {
    self.trailers.removeValue(forKey: name)
    return self
  }

  @discardableResult
  public func removeTrailers(named name: String, value: String) -> ResponseBuilder {
    self.trailers[name]?.removeAll(where: { $0 == value })
    if self.trailers[name]?.isEmpty == true {
      self.trailers.removeValue(forKey: name)
    }

    return self
  }

  @discardableResult
  public func addBody(_ body: Data?) -> ResponseBuilder {
    self.body = body
    return self
  }

  public func build() -> Response {
    return Response(statusCode: self.statusCode,
                    headers: self.headers,
                    trailers: self.trailers,
                    body: self.body)
  }
}

// MARK: - Objective-C helpers

extension Response {
  /// Convenience builder function to allow for cleaner Objective-C syntax.
  ///
  /// For example:
  ///
  /// Response *res = [Response withBuild:^(ResponseBuilder *builder) {
  ///   [builder addBody:bodyData];
  ///   [builder addHeaderWithName:@"x-some-header" value:@"foo"];
  ///   [builder addTrailerWithName:@"x-some-trailer" value:@"foo"];
  /// }];
  @objc
  public static func with(build: (ResponseBuilder) -> Void) -> Response {
    let builder = ResponseBuilder()
    build(builder)
    return builder.build()
  }
}
